Hold Pod in cache until all other cleanup work is completed

Huang-Wei opened the pull request:
**What type of PR is this?**

/kind flake
/sig scheduling
/priority critical-urgent

**What this PR does / why we need it**:

The PreBind runs in a goroutine, and unfortunately, the `assumedPod` gets removed from cache **prior to** clean-up work (e.g., Unreserve) gets finished:

https://github.com/kubernetes/kubernetes/blob/498b1e227981e1fe317d91143617436e81c27775/pkg/scheduler/scheduler.go#L577-L588

This causes tricky flakes. The flaky test was built on a buffered channel (len=3) to receive Unreserve events. However, it's possible that when the binding goroutine of the previous scheduling attempt is waiting for Unreserve to finish (L585), another scheduling attempt starts, and it proceeds to also run Unreserve. The 2 Unreserve running together will compete to fill in the buffered channel (len=3), which causes L586 (above) in either goroutines unreachable. And in the [flake log], there are 32 binding goroutines running at the same time...

This PR tries to fix the problem by:

- Move "ForgetPod" after "RunReservePluginsUnreserve", so that the cache would hold the pod to avoid it's being retried simutaneously until Unreserve is completed.
- Move "assume" ahead of "RunReservePluginsReserve". This is based on the fact that "ForgetPod" is the last step of failure path, so "assume" should be reversly treated as the first step. The current failure path is like this:
    ```
    assume -> reserve -> unreserve -> forgetPod -> recordingFailure
    ``` 
- Make subtests of `TestReservePluginUnreserve` stateless

**Which issue(s) this PR fixes**:

Fixes TestReservePluginUnreserve flake in #93485.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

[flake log]: https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-integration-master/1287660403844714499

Huang-Wei commented:
/assign @ahg-g 
/cc @adtac @alculquicondor

alculquicondor commented:
if we go with the logic that "assume" is a reserve operation, we should go: reserve->assume->forget->unreserve (like in master)
or: assume->reserve->unreserve->forget (another option that we could consider, but not sure if it works well with volume binding).

Recall that we run the Reserve plugins's Unreserve in the opposite order.

Isn't the bug just in the integration test implementation? We shouldn't be relying in a single channel.

alculquicondor commented:
cc @cofyc

Huang-Wei commented:
> or: assume->reserve->unreserve->forget (another option that we could consider, but not sure if it works well with volume binding).

This sounds good to me - if we think of "assume" as a special "reserve". @cofyc WDYT?

> Recall that we run the Reserve plugins's Unreserve calls in the opposite order.

Correct. I didn't touch that piece.

> Isn't the bug just in the integration test implementation? We shouldn't be relying in a single channel.

Yes or no. The users may also make similar assumptions I guess?

adtac commented:
What I don't fully understand is, why is another scheduler attempt being made when there's no indication that the existing schedule attempt is complete? Perhaps I've misunderstood something, but isn't `sched.recordSchedulingFailure` used to let the scheduler know that a pod is unschedulable so that it can be put back into the queue? If so, why would another attempt begin for this pod before `sched.recordSchedulingFailure` is called from within the goroutine?

alculquicondor commented:
Is it a single pod being scheduled?

adtac commented:
Yes.

alculquicondor commented:
Then different pods' scheduling attempts are competing for the same channel. Recall that binding is a separate go routine.

alculquicondor commented:
synced with @adtac offline. He's going to investigate the fix in the test.

I fear that changing the order in which reserve and assume happen might trigger other problems. Plugins should be more careful in their implementation.

Huang-Wei commented:
> Perhaps I've misunderstood something, but isn't sched.recordSchedulingFailure used to let the scheduler know that a pod is unschedulable so that it can be put back into the queue?

It's because in the current logic, the "assumedPod" gets removed (`ForgetPod`) from cache first, and then it proceeds to run Unreserve, so during that, if the Pod gets updated somehow (in the test, it's because of the forced recync. We can disable that, but I guess it's just mitigation), we don't have the cache entry to block the update-event trggered scheduling attempt. So the 2nd attempt would compete with the original ongoing Unserve on the shared buffer-channel, and both will hang due to the limited capacity.

Huang-Wei commented:
> I fear that changing the order in which reserve and assume happen might trigger other problems. Plugins should be more careful in their implementation.

@alculquicondor Do you want me to append a commit to move "assumePod" prior to "RunReservePluginsReserve"?

alculquicondor commented:
I'm more in favor of keeping the current behavior. At least until we get some input from Yecheng.

adtac commented:
Two points:

1. creating a separate event channel for each `t.Run` invocation is probably good practice even though it won't help with this flaky test.

2. @Huang-Wei can you explain why the order of recordSchedulingFailure and RunReservePluginsUnreserve is reversed compared to master? I understand why the pair of them must occur before the cache forget operation, but why is the order within the pair reversed?

alculquicondor commented:
To be more specific, @cofyc do you see any problems with the following order:

```assume->reserve->unreserve->forget```

cofyc commented:
> > or: assume->reserve->unreserve->forget (another option that we could consider, but not sure if it works well with volume binding).
> 
> This sounds good to me - if we think of "assume" as a special "reserve". @cofyc WDYT?
> 

it should not have any problem with volume binding.

IIUC, in asynchronous binding goroutine, if the pod is forgotten by the scheduler cache, then it can be retried by the future scheduler cycles?

Then I think this is better compared to current behavior. The pod should be released (forgot in scheduler cache) after all cleanups are finished.

cofyc commented:
> To be more specific, @cofyc do you see any problems with the following order:
> 
> `assume->reserve->unreserve->forget`

I don't see any problems for now.

Huang-Wei commented:
> 1. creating a separate event channel for each t.Run invocation is probably good practice even though it won't help with this flaky test.

That needs to be updated for sure, but not key to fix the issue. We can come up with a general PR fixing all the non-stateless tests.

> 2. @Huang-Wei can you explain why the order of recordSchedulingFailure and RunReservePluginsUnreserve is reversed compared to master? I understand why the pair of them must occur before the cache forget operation, but why is the order within the pair reversed?

As I explained in the description section, it's optional. I just felt that given `recordSchedulingFailure` is already a fact, and running Unreserve may take time, so it makes sense to move it ahead of Unreserve, which can make other actors that rely on a unschedulable status act fast accordingly. And also, some existing paths such as reserve<->unreserve and assume<->unreserve is already putting `recordSchedulingFailure` ahead of unreserve.
But again, it's optional.

Huang-Wei commented:
> IIUC, in asynchronous binding goroutine, if the pod is forgotten by the scheduler cache, then it can be retried by the future scheduler cycles?

Yes. In any failure path, the failed Pod will be pathed with an (updated) unschedualbe status. So that will trigger the failed pod to be re-enqueued.

alculquicondor commented:
So `recordSchedulingFailure` should happen at the end.

Are you sure the updated scheduled status causes the pod to be put back into active?

Huang-Wei commented:
@adtac Due to https://github.com/kubernetes/kubernetes/pull/93511#issuecomment-665316132, I think we still need to put `recordSchedulingFailure` after `ForgetPod`. So I'd propose the order to be:

- Happy path: assume ->reserve->unreserve->forget
- Negative path: assume->reserve->unreserve->forget->recordSchedulingFailure

@alculquicondor @cofyc WDYT?

Huang-Wei commented:
> Are you sure the updated scheduled status causes the pod to be put back into active?

Not necessarily activeQ, but should be queued anyhow:

https://github.com/kubernetes/kubernetes/blob/22329fbe1338934d5482d91c5dec18851c7ddeaf/pkg/scheduler/eventhandlers.go#L264

alculquicondor commented:
> Not necessarily activeQ, but should be queued anyhow:

The pod is put back into unschedulable queue at the end of the scheduling/binding cycles in case of error as well.

> Happy path: assume ->reserve->unreserve->forget

I thought unreserve and forget is only called in the unhappy path.

Huang-Wei commented:
> The pod is put back into unschedulable queue at the end of the scheduling/binding cycles in case of error as well.

Yes, I appended a commit to keep recordingFailure at the last step of failure path.

> I thought unreserve and forget are only called in the unhappy path.

Yes, updated the wordings.

Huang-Wei commented:
> Yes, I appended a commit to keep recordingFailure at the last step of failure path.

@alculquicondor @adtac @cofyc PTAL.

cofyc commented:
> * r

SGTM

cofyc commented:
> @adtac Due to [#93511 (comment)](https://github.com/kubernetes/kubernetes/pull/93511#issuecomment-665316132), I think we still need to put `recordSchedulingFailure` after `ForgetPod`. So I'd propose the order to be:
> 
> * assume->reserve->unreserve->forget->recordSchedulingFailure.
> 
> @alculquicondor @cofyc WDYT?

SGTM

cofyc commented:
https://github.com/kubernetes/kubernetes/blob/22329fbe1338934d5482d91c5dec18851c7ddeaf/pkg/scheduler/scheduler.go#L590-L595

when the binding operation fails, the `unreserve` functions are called after the `ForgetPod` (called in the `sched.bind`). should we update this too?

cofyc commented:
https://github.com/kubernetes/kubernetes/blob/22329fbe1338934d5482d91c5dec18851c7ddeaf/pkg/scheduler/scheduler.go#L505-L527

I guess `sched.assume` must be called before calling `reserve` functions. and no `unreserve` functions need to be called when it fails.

`assume/forget` works like a special `reserve` plugin which is always the first in the plugin list.

Huang-Wei commented:
/retest

Huang-Wei commented:
> I guess `sched.assume` must be called before calling `reserve` functions. and no `unreserve` functions need to be called when it fails.

@cofyc that makes sense to me. But maybe in a separate PR, and companies with a test.

Huang-Wei commented:
/retest

alculquicondor commented:
PR description is still outdated.

alculquicondor commented:
/hold
/lgtm

alculquicondor commented:
/lgtm cancel

Let's do assume->reserve

Huang-Wei commented:
@alculquicondor @adtac Please take a look at the latest 2 commits:

- one commit moves assume() ahead of RunReservePluginsReserve()
- the other commit makes subtests stateless. Although it increases the overall running time a bit, it still worth it IMO, as based on the updated logic, the **shared** buffered channel may have already been re-filled up (due to retried goroutines) when a later subtest starts

@ahg-g would you mind adding it to 1.19 release? I will squash later.

alculquicondor commented on test/integration/scheduler/framework_test.go:
```diff
@@ -1080,12 +1080,11 @@ func TestReservePluginUnreserve(t *testing.T) {
-				pl.reset()
-			}
-			testutils.CleanupPods(testCtx.ClientSet, t, []*v1.Pod{pod})
+			// preBindPlugin.reset()
```
Is this comment intentional?

adtac replied:
Should probably be removed since each set of plugins is used only once.

Huang-Wei replied:
Yes, reset() are not necessary as each subtest is using its own plugin.

alculquicondor reviewed (commented):
code lgtm. @adtac, PTAL at the updated test.

adtac commented on test/integration/scheduler/framework_test.go:
```diff
@@ -1080,12 +1080,11 @@ func TestReservePluginUnreserve(t *testing.T) {
+			// for _, pl := range reservePlugins {
+			// 	pl.reset()
+			// }
+			// testutils.CleanupPods(testCtx.ClientSet, t, []*v1.Pod{pod})
```
This shouldn't be commented out, however.

Huang-Wei replied:
Yes, it should be kept.

adtac commented:
IIRC there were other integration tests that shared a registry, KubeSchedulerProfile, etc. too; I followed the pattern in those tests when writing TestReservePluginUnreserve. While the shared channel is the primary source of test failures, not having ANY shared state might be a good idea in other tests too.

Huang-Wei committed:
Hold Pod in cache until all other cleanup work is completed

- Move "ForgetPod" after "RunReservePluginsUnreserve", so that the cache would hold the pod to
  avoid it's being retried simutaneously until Unreserve is completed.
- Move "assume" ahead of "RunReservePluginsReserve". This is based on the fact that "ForgetPod" is
  the last step of failure path, so "assume" should be reversly treated as the first step. The
  current failure path is like this:
  assume -> reserve -> unreserve -> forgetPod -> recordingFailure
- Make subtests of TestReservePluginUnreserve stateless

Huang-Wei commented:
@adtac It'd much appreciate if you could craft a followup PR to clean up those tests that hold shared state.

alculquicondor reviewed (commented):
/lgtm

adtac commented:
/lgtm

(but I don't think my /lgtm does anything)

Huang-Wei commented:
/hold

Need to take a look at the volumebinding failure: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/93511/pull-kubernetes-integration/1288528886769913856

adtac commented:
>It'd much appreciate if you could craft a followup PR to clean up those tests that hold shared state.

Will try to do that :+1:

Huang-Wei commented:
Listing records of TestVolumeBinding:

- Pass: [P1] 
- Failure: [F1]

[P1]: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/93511/pull-kubernetes-integration/1288542065629597697
[F1]: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/93511/pull-kubernetes-integration/1288528886769913856

Huang-Wei commented:
/test pull-kubernetes-integration
/test pull-kubernetes-bazel-test

ahg-g commented:
/milestone v1.19

Huang-Wei commented:
/retest

Huang-Wei commented:
/test pull-kubernetes-integration

Huang-Wei commented:
/retest

Huang-Wei commented:
/test pull-kubernetes-integration

Huang-Wei commented:
/test pull-kubernetes-integration

Huang-Wei commented:
/test pull-kubernetes-integration

cofyc commented:
> https://github.com/kubernetes/kubernetes/blob/22329fbe1338934d5482d91c5dec18851c7ddeaf/pkg/scheduler/scheduler.go#L590-L595
> 
> when the binding operation fails, the `unreserve` functions are called after the `ForgetPod` (called in the `sched.bind`). should we update this too?

just a reminder that we should run `unreserve` functions before the `ForgetPod` when the binding operation fails for consistency. refactor in a follow-up PR?

Huang-Wei commented:
> just a reminder that we should run `unreserve` functions before the `ForgetPod` when the binding operation fails for consistency. refactor in a follow-up PR?

Yes, we should do that. Except for that, do you think the failures in https://github.com/kubernetes/kubernetes/pull/93511#issuecomment-665861157 is related to the new changes of this PR?

cofyc commented:
> > just a reminder that we should run `unreserve` functions before the `ForgetPod` when the binding operation fails for consistency. refactor in a follow-up PR?
> 
> Yes, we should do that. Apart from that, do you think the failures in [#93511 (comment)](https://github.com/kubernetes/kubernetes/pull/93511#issuecomment-665861157) is related to the new changes of this PR?

I don't think so. Similar failures happen in master and other PRs too. https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&test=TestVolumeProvision

I'll track this flaky test in https://github.com/kubernetes/kubernetes/issues/81368.

Huang-Wei commented:
Thanks @cofyc ! How about `TestVolumeBinding`, I observed 2 failures (F1 & F3) so far in https://github.com/kubernetes/kubernetes/pull/93511#issuecomment-665861157.

cofyc commented:
> Thanks @cofyc ! How about `TestVolumeBinding`, I observed 2 failures (F1 & F3) so far in [#93511 (comment)](https://github.com/kubernetes/kubernetes/pull/93511#issuecomment-665861157).

I think it's unrelated.

Huang-Wei commented:
/test pull-kubernetes-integration

Huang-Wei commented:
/test pull-kubernetes-integration

Huang-Wei commented:
> I think it's unrelated.

Agree. I spot it [flaked](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/93550/pull-kubernetes-integration/1288687271234506755) in an irrelevant PR (#93550)

So
/hold cancel
